### DISCUSSION TITLE
CVE-2021-26910 and `--overlay-tmpfs`

### DISCUSSION THREAD

**Discussion by laniakea64** [netblue30/firejail#4178] [Q&A] [answered]
I know that as an emergency fix for this CVE, firejail disabled all the overlay-related features, "fixes pending" - https://github.com/netblue30/firejail/commit/fb9f2a5fb3ac1ebbb14302ecdf3c840b70b090da

As I depend on extensive use of `--overlay-tmpfs`, and this is still disabled / fixes still pending, this became very concerning.
Sorry I'm a bit late asking about this (been busy), but looking at and testing with the publicly available exploit code, I gather that:

- Exploiting this vulnerability is not done from inside the firejail sandbox.  It is done by a malicious program invoking firejail from outside a sandbox.

- The vulnerability is specifically with the overlay features that mount (persistent) overlays inside the home directory. (which I don't use)

Is this assessment correct?

Also I have two questions:

1. On my Xubuntu 20.04 + firejail 0.9.64, the exploit code fails:

```console
$ gcc -g -fPIC -o UnjailMyHeart UnjailMyHeart.c
$ ./UnjailMyHeart
Using default firejail startup arguments.
Before sleep.
Failed to rename directories 2 (No such file or directory).
Renamed ".firejail" directory.
Exploitation failed.
```

And these overlay features are broken anyway:

```console
$ firejail --overlay --noprofile
Parent pid 4040, child pid 4042

**     Warning: dropping all Linux capabilities     **

Debug: running on kernel version 5.4
Error mounting overlayfs: fs.c:1063 fs_overlayfs: Too many levels of symbolic links
Error: proc 4040 cannot sync with peer: unexpected EOF
Peer 4042 unexpectedly exited with status 1
```

Does it mean my setup is not vulnerable?

```console
$ firejail --version
firejail version 0.9.64

Compile time support:
        - AppArmor support is enabled
        - AppImage support is enabled
        - chroot support is enabled
        - D-BUS proxy support is enabled
        - file and directory whitelisting support is enabled
        - file transfer support is enabled
        - firetunnel support is disabled
        - networking support is enabled
        - overlayfs support is enabled
        - private-home support is enabled
        - SELinux support is disabled
        - user namespace support is enabled
        - X11 sandboxing support is enabled
```

2. As `--overlay-tmpfs` mounts in a firejail-controlled, root-owned subdirectory of `/run`, is it immune to this vuln?  If so, could you please restore the `--overlay-tmpfs` while keeping the other, still-vulnerable overlay features disabled (and thus keeping this vulnerability closed)?

Thanks!

---

Relates to:

* CVE-2021-26910
* #2799
* #5658
* #6041
* #6632

**Comment by reinerh** [comment]
Ping @netblue30.

I think one of the reasons to just disable `--overlay` was that it is broken on "newer" kernels anyway, as you already noticed.

I don't know why it fails on your Xubuntu, which still has a vulnerable version. Maybe the exploit code is not flexible enough.

**Reply by laniakea64** [reply]
> Ping @netblue30.

Hi @netblue30 have you been able to look at this yet?  Thanks

**Comment by laniakea64** [comment]
Just happened on https://github.com/netblue30/firejail/issues/4006#issuecomment-785154579 -
> You simply test if `firejail --overlayfs-named=foobar --noprofile bash` works. If it works, you're vulnerable.

Does this mean the answer to my question (1) is I'm not vulnerable?

**Reply by rusty-snake** [reply]
This command fails if (a) firejail is build w/o overlayfs support, (b) overlayfs support is disable in firejail.config or (c) you use a newer kenrel where firejail overlayfs implementation is broken. In all three cases you can not use `--overlayfs`/`--overlayfs-named` and therefore it can not be exploited (as I understand everything).

FYI:  #2799

**Comment by laniakea64** [comment]
> As `--overlay-tmpfs` mounts in a firejail-controlled, root-owned subdirectory of `/run`, is it immune to this vuln?  If so, could you please restore the `--overlay-tmpfs` while keeping the other, still-vulnerable overlay features disabled (and thus keeping this vulnerability closed)?

Hi, with 0.9.66 now on the way to release, it would be awesome to have an answer to this, especially if `--overlay-tmpfs` alone can be safely restored.  Thanks!

**Comment by laniakea64** [comment]
It seems @netblue30 is too busy to even respond to this, and no one is contradicting https://github.com/netblue30/firejail/discussions/4178#discussioncomment-771979 which basically says `overlay-tmpfs` just got thrown out with the bathwater and is not itself vulnerable.  So could propose the following resolution? -

 - Completely separate build-time options for `overlayfs` and `overlay-tmpfs`.  In all the time I've used firejail, support/usability of these two features has always been _very_ different, so this would seem to make sense anyway.
 - Keep the build-time option for `overlayfs` hard-disabled / commented out until that feature is really fixed.
 - For the moment, make the build-time option for `overlay-tmpfs` disabled by default, but still accessible for those who need it (like the apparmor option is now).

Would this be reasonable?

**Reply by smitsohu** [reply]
If I recall correctly, the `overlay-tmpfs` option is indeed unaffected.

To me your proposal sounds reasonable, but it would be a stopgap only.

**Reply by laniakea64** [reply]
Thanks smitsohu :slightly_smiling_face: 

Any update on this?

**Comment by laniakea64** [comment]
Hi, any news on this?

Unfortunately I don't have enough C knowledge to pull request the proposal in https://github.com/netblue30/firejail/discussions/4178#discussioncomment-1387099 myself.  Could this be done before or soon after the upcoming 0.9.68 release?

If not, since my particular systems are not affected by the vulnerability, will I need to just manually revert that part of https://github.com/netblue30/firejail/commit/fb9f2a5fb3ac1ebbb14302ecdf3c840b70b090da for my local builds for now to update firejail beyond 0.9.64 without losing the important `--overlay-tmpfs`?

**Reply by rusty-snake** [reply]
>  If not, since my particular systems are not affected by the vulnerability, will I need to just manually revert that part of fb9f2a5 for my local builds for now to update firejail beyond 0.9.64 without losing the important --overlay-tmpfs?

Yes, `-DHAVE_OVERLAYFS` should enable this code again.

**Reply by laniakea64** [reply]
So, this had been working well until one of my systems was upgraded to kernel 5.19.0.  Apparently this kernel update put `--overlay-tmpfs` on the wrong side of a broken kernel version check intended to disable `--overlay` under kernels 4.19 and later - https://github.com/netblue30/firejail/issues/2799#issuecomment-551891564

So now I'm experimenting with the following patch in addition, which testing it seems to successfully restore `--overlay-tmpfs` under my 5.19.0 kernel while keeping `--overlay` blocked as intended -
```diff
diff --git a/src/firejail/fs_overlayfs.c b/src/firejail/fs_overlayfs.c
index 167a7e28b..375ddbdf5 100644
--- a/src/firejail/fs_overlayfs.c
+++ b/src/firejail/fs_overlayfs.c
@@ -144,8 +144,8 @@ void fs_overlayfs(void) {
 
        // mounting an overlayfs on top of / seems to be broken for kernels > 4.19
        // we disable overlayfs for now, pending fixing
-       if (major >= 4 &&minor >= 19) {
-               fprintf(stderr, "Error: OverlayFS disabled for Linux kernels 4.19 and newer, pending fixing.\n");
+       if (arg_overlay_keep) {
+               fprintf(stderr, "Error: OverlayFS disabled, pending fixing.\n");
                exit(1);
        }
 
```
(And I wonder if this would also completely eliminate the vulnerability?)

**Comment by RaphaelJenni** [comment]
Is there any news regarding a fix for the overlays feature in firejail. I need a tool that lets me run processes as if they were on the host machine but captures all changes that were made to the filesystem in a different folder.

Firejail seemed to be able to do this until the option got remvoed.

**Comment by illtellyoulater** [comment]
@netblue30 any news on this?

**Reply by kmk3** [reply]
@illtellyoulater [on Apr 27](https://github.com/netblue30/firejail/discussions/4178#discussioncomment-5741995):

> any news on this?

If there was, it would likely have already been posted in here.

If you have something useful to contribute (such as fixing the underlying
security vulnerability or showing that --overlay-tmpfs would not be affected by
it anymore), then feel free to do so.

Otherwise, please stop spamming everyone.

**Comment by illtellyoulater** [comment]
> [Otherwise, please stop spamming everyone.](https://github.com/netblue30/firejail/discussions/4178#discussioncomment-5742175)

@kmk3  

Ok... Listen up.
This thread was opened 2 years ago, and despite the various feedback provided by @laniakea64 and the practical proposals he made to bring back at least a safe subset of overlay-tmpfs, he eventually just ended up being ignored.... None of you cared to speak your mind, showing no interest in making us aware of your thoughts... 
The only notable exception was @smitsohu who initially said to be favorable but later vanished too...

After another year, same thing happened again with @RaphaelJenni's question above from October 2022, which was also left unanswered.

Now, after another 6 months of silence, on the very second I ask for updates (as a new user and yes, in no less than 2 different threads, after seeing replies stuck at 2021), you come up and try to pass me off as an annoying spammer?
Are you ok man?? Perhaps you think you are the only busy person here? 
Or that your time is more valuable than anyone else's? 
Unbelievable...

It would be much easier if you guys would let us know once and for all, what's your thoughts and how you intend to proceed about this approach of just reintroducing the safe parts of overlay code.

In case you can't be bothered answering such a simple question and intend to leave things as they are (no overlay support) then stop being the first root of all confusion by at least removing `--overlay-tmpfs` [documentation from Firejail website](https://firejail.wordpress.com/documentation-2/basic-usage/#overlayfs).

Either way, don't you dare pegging people as spammers solely as a result of your negligence.

**Comment by kmk3** [comment]
@illtellyoulater [on Apr 28](https://github.com/netblue30/firejail/discussions/4178#discussioncomment-5753461):

> > [Otherwise, please stop spamming
> > everyone.](https://github.com/netblue30/firejail/discussions/4178#discussioncomment-5742175)
> 
> @kmk3
> 
> Ok... Listen up. This thread was opened 2 years ago, and despite the various
> feedback provided by @laniakea64 and the practical proposals he made to bring
> back at least a safe subset of overlay-tmpfs, he eventually just ended up
> being ignored....
> 
> None of you cared to speak your mind, showing no interest in making us aware
> of your thoughts... The only notable exception was @smitsohu who initially
> said to be favorable but later vanished too...
> 
> After another year, same thing happened again with @RaphaelJenni's question
> above from October 2022, which was also left unanswered.
> 
> Now, after another 6 months of silence, on the very second I ask for updates
> (as a new user and yes, in no less than 2 different threads, after seeing
> replies stuck at 2021)

Which again, likely means that nothing has changed.

Given that no replies were posted, it's likely that the developers that are
familiar with the feature don't have the time to work on/review it, are not
interested in working on/reviewing it and/or are more interested in working in
something else instead.

If nobody qualified to do the work is available to do it, then it follows that
the work won't be done.

> you come up and try to pass me off as an annoying spammer? Are you ok man??

To go and "ask for updates" / ETAs on the bug trackers and similar tools of
free software projects only adds noise; it contributes nothing to the issue at
hand.  So yes, that would generally be considered spamming.  Note also that
there are 6 other people in this thread and that likely more are subscribed to
this discussion and/or to the project in general; posting such comments only
serves to send useless notifications to everyone.

If you are so passionate about this feature, why haven't you stepped up yet to
audit and fix it?  You are free to get familiar with the code and with the
issues and to find a way to fix them.

You are also free to fork the project and add the feature back yourself.

> Perhaps you think you are the only busy person here? Or that your time is
> more valuable than anyone else's? Unbelievable...

Given your demeanor, I think you're just projecting.

> It would be much easier if you guys would let us know once and for all,
> what's your thoughts and how you intend to proceed about this approach of
> just reintroducing the safe parts of overlay code.

I think it would make sense for those who are interested in this feature to go
test and audit the existing overlayfs code to make sure that it's working
properly.  Then, for any straightforward fixes and improvements, feel free to
open PRs.  For concrete non-trivial proposals, it would be a good idea to first
open an issue about it and reference it in here.

> In case you can't be bothered answering such a simple question and intend to
> leave things as they are (no overlay support) then stop being the first root
> of all confusion by at least removing `--overlay-tmpfs` [documentation from
> Firejail website](https://firejail.wordpress.com/documentation-2/basic-usage/#overlayfs).

If it's still on the website, then that's a bug in it.

Though note that the website is probably not updated often; the man pages are
the official documentation and `--overlay-tmpfs` is not listed in them.

> Either way, don't you dare pegging people as spammers solely as a result of
> your negligence.

That's a funny one.  By proclaiming "negligence" you seem to be under the
impression that you are entitled to free labor from other people.

In which case, what about _your_ negligence?  Why haven't you put in the work
to bring back the feature yet?  As you said it yourself, it's already been 2
years.  Got any updates to share on it?

**Reply by laniakea64** [reply]
> To go and "ask for updates" / ETAs on the bug trackers and similar tools of free software projects only adds noise; it contributes nothing to the issue at hand. So yes, that would generally be considered spamming.

Hi @kmk3 , unless this person asked the same question on multiple threads, the usual definitions of "spamming" don't go that far.  If you want to use this extended definition of spamming, it's your choice, but in that case please do the firejail community a favor and codify it in a project Code of Conduct or similar.  Otherwise, people _will_ unwittingly "spam" in this way, and when called on it, will be caught by surprise and take it as unwarranted aggression, as appears to have happened here.  Thanks.

> If nobody qualified to do the work is available to do it, then it follows that the work won't be done.
> ...
> I think it would make sense for those who are interested in this feature to go test and audit the existing overlayfs code to make sure that it's working properly. Then, for any straightforward fixes and improvements, feel free to open PRs. For concrete non-trivial proposals, it would be a good idea to first open an issue about it

+1, and please lock this discussion, since that means there is nothing else constructive for a non-collaborator to say in this discussion atm.  And I no longer have the capacity to deal with this subject anymore, especially in light of the last few comments here.  I'm sorry for asking for firejail to change in this area, I had no idea that was a tense topic, nor how tough it would be for firejail devs to be able to address this officially in firejail code.

**Reply by kmk3** [reply]
@laniakea64 [on Apr 29](https://github.com/netblue30/firejail/discussions/4178#discussioncomment-5758269):

> > To go and "ask for updates" / ETAs on the bug trackers and similar tools of
> > free software projects only adds noise; it contributes nothing to the issue
> > at hand. So yes, that would generally be considered spamming.
> 
> Hi @kmk3 , unless this person asked the same question on multiple threads,
> the usual definitions of "spamming" don't go that far.
> 
> Otherwise, people _will_ unwittingly "spam" in this way, and when called on
> it, will be caught by surprise and take it as unwarranted aggression, as
> appears to have happened here. Thanks.

My point is that simply bumping a thread without engaging in the conversation
in any way is not productive.  Maybe other projects on GitHub treat the
"Discussions" section differently, but I'd argue that posting such comments is
generally frowned upon in development-focused spaces (as opposed to places like
online forums).

As for the wording, "spamming" was not meant as a personal attack and I
honestly did not anticipate that it could become a point of contention.

I meant it in the sense of someone sending one or more unsolicited/unproductive
messages to multiple people (as in the original/classic spam email and such).

That is, even if it was not the intention, posting a single comment might
result in an email/notification being sent to dozens of people.

In my mind the intent seemed clear, but from the responses it could have been
worded better/more clearly.

Anyway, sorry everyone for any grievances that this might have caused.

> If you want to use this extended definition of spamming, it's your choice,
> but in that case please do the firejail community a favor and codify it in a
> project Code of Conduct or similar.

I suppose that it would be helpful to have something written down somewhere, at
least so that it can be referred to later.  Might open a PR at some point to
add some text to CONTRIBUTING.md.

> And I no longer have the capacity to deal with this subject anymore,
> especially in light of the last few comments here.

That's unfortunate but understandable.  Personally, I remember your comments
generally being helpful in other discussions in the project as well, so I'd
look forward to any contributions that you might make, regardless of the topic.

> I'm sorry for asking for firejail to change in this area, I had no idea that
> was a tense topic, nor how tough it would be for firejail devs to be able to
> address this officially in firejail code.

At a glance the problem seems non-trivial, but I don't think that it is
necessarily a tense topic.  Though keep in mind that any given topic that tends
towards a low signal to noise ratio is likely to also end up driving away
people that might otherwise be interested in contributing to it (especially if
the noise reaches their inboxes).

To be clear, I was talking about posts that essentially amount to a thread
bump.  [This post][1] at least mentions a use case that wasn't mentioned
before, so it is constructive in that regard.

I see that you also have asked for updates in a few comments, but to your
credit, in this thread you have also posed valid questions, made reasonable
proposals and attempted to fix/work around the problem in one way or another
(and if it helps, I didn't really have you in mind when writing the previous
comment).

This is probably why I hadn't said anything before, but in fairness, it would
probably have been better to have done so (again, worded in a different
manner).

> > If nobody qualified to do the work is available to do it, then it follows
> > that the work won't be done.  ...  I think it would make sense for those
> > who are interested in this feature to go test and audit the existing
> > overlayfs code to make sure that it's working properly. Then, for any
> > straightforward fixes and improvements, feel free to open PRs. For concrete
> > non-trivial proposals, it would be a good idea to first open an issue about
> > it
> 
> +1, and please lock this discussion, since that means there is nothing else
> constructive for a non-collaborator to say in this discussion atm.

I think I'll leave this as is for a few days and then hide all of these recent
comments and add a new comment asking to avoid posting +1 comments and things
like that.  I'll try to word it more positively.

Then I suppose you're right and that it would make sense to lock it, as any
feature with a low ratio of maintainership to popularity seems likely to
attract unproductive discussion.

Maybe it would be worth it for some individual suggestions to be turned into
their own dedicated discussions.  Or, for each more straightforward and/or more
clearly defined proposals (such as the one about having separate configure
flags for --overlay and --overlay-tmpfs), a [feature request][2] could be
opened.

[1]: https://github.com/netblue30/firejail/discussions/4178#discussioncomment-3933031
[2]: https://github.com/netblue30/firejail/issues/new?template=feature_request.md

**Comment by rusty-snake** [comment]
https://github.com/netblue30/firejail/pull/6632#discussion_r1929573432

>   * crablock and bubblewrap have integrated support for overlayfs. This was possible because Linux >= 5.11 allows `overlayfs` mounts from unprivileged user-namespace. An "unprivileged" implementation in firejail might be the most likely comeback as it means less auditing.

**Comment by rusty-snake** [comment]
Linking #6994 in case anyone is watching this issue.

**Comment by kmk3** [comment, marked as answer]
Closing, as overlayfs support was removed:

* #6994
